GVRf/Framework/jni/shaders/material/custom_shader.cpp

```diff
@@ -29,91 +29,213 @@
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformVec2Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_vec2_keys_[location] = key;
+void CustomShader::addUniformVec2Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::vec2 v = material.getVec2(key);
+                glUniform2f(location, v.x, v.y);
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformVec3Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_vec3_keys_[location] = key;
+void CustomShader::addUniformVec3Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::vec3 v = material.getVec3(key);
+                glUniform3f(location, v.x, v.y, v.z);
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformVec4Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_vec4_keys_[location] = key;
+void CustomShader::addUniformVec4Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::vec4 v = material.getVec4(key);
+                glUniform4f(location, v.x, v.y, v.z, v.w);
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformMat4Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_mat4_keys_[location] = key;
+void CustomShader::addUniformMat4Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::mat4 m = material.getMat4(key);
+                glUniformMatrix4fv(location, 1, GL_FALSE, glm::value_ptr(m));
+            };
+    addUniformKey(variable_name, key, f);
 }
 
 void CustomShader::render(const ShaderUniformsPerObject& uniforms, RenderData* render_data,
         const std::vector<Light*> lightList, Material* material) {
-    for (auto it = texture_keys_.begin(); it != texture_keys_.end(); ++it) {
-        Texture* texture = material->getTextureNoError(it->second);
-        // If any texture is not ready, do not render the material at all
-        if (texture == NULL || !texture->isReady()) {
-             return;
+    initializeOnDemand();
+
+    {
+        std::lock_guard<std::mutex> lock(textureVariablesLock_);
```

Comment:
We need to make somehow the access to the std collection thread-safe. This change was the most conservative. We may even think about eliminating the loop itself and/or employing a copy-on-write scheme for the collection(s), etc.

The very same loop takes wall clock time in the range of 0.0015-0.003ms without locking and in the range of 0.006-0.02ms with the locking so definitely there is overhead, especially in the worst case. But I find this matter to be applicable to any place we use locks in the framework - we must ensure correctness first and worry about performance once a bottleneck is identified.

--

If the check is in f_bind then you can end up with some textures bound and some not; the current behavior is that if any texture is not ready we short-circuit the rendering so I am not sure the check can be moved to f_bind.

-

Let me know how big of a problem is this for you and is it something you want tackled as part of this pull request. I probably would try a copy-on-write variant where I would push the overhead to the mutator methods that presumably are not on the gl thread. And they will not be in the near future.